lib/3scale/backend/stats/cleaner.rb

```diff
@@ -0,0 +1,156 @@
+        # https://github.com/3scale/apisonator/issues/90
+
+        include Storable
+
+        SET_WITH_SERVICES_MARKED_FOR_DELETION = 'set_with_services_marked_for_deletion'.freeze
+        private_constant :SET_WITH_SERVICES_MARKED_FOR_DELETION
+
+        SLEEP_BETWEEN_SCANS = 0.01 # In seconds
+        private_constant :SLEEP_BETWEEN_SCANS
+
+        SCAN_SLICE = 500
+        private_constant :SCAN_SLICE
+
+        STATS_KEY_PREFIX = 'stats/'.freeze
+        private_constant :STATS_KEY_PREFIX
+
+        class << self
+          include Logging
+          def mark_service_to_be_deleted(service_id)
+            storage.sadd(SET_WITH_SERVICES_MARKED_FOR_DELETION, service_id)
+          end
+
+          # Deletes all the stats for the services that have been marked for
+          # deletion.
+          #
+          # This method receives a collection of instantiated Redis clients.
+          # Those clients need to connect to Redis servers directly. They cannot
+          # connect to a proxy like Twemproxy. The reason is that this function
+          # needs to scan the database using the "SCAN" command, which is not
+          # supported by Twemproxy.
+          #
+          # The services marked as deletion will be marked as done only when
+          # this function finishes deleting the keys from all the Redis servers.
+          # This means that if the function raises in the middle of the
+          # execution, those services will be retried in the next call.
+          #
+          # Note 1: keys deleted cannot be restored.
+          # Note 2: this method can take a long time to finish as it needs to
+          # scan all the keys in several DBs.
+          def delete!(redis_conns, log_deleted_keys: false)
+            services = services_to_delete
+            logger.info("Going to delete the stats keys for these services: #{services.to_a}")
+
+            redis_conns.each do |redis_conn|
+              delete_keys(redis_conn, services, log_deleted_keys)
+            end
+
+            remove_services_from_delete_set(services)
+          end
+
+          private
+
+          # Returns a set with the services included in the
+          # SET_WITH_SERVICES_MARKED_FOR_DELETION Redis set.
+          def services_to_delete
+            res = []
+            cursor = 0
+
+            loop do
+              cursor, services = storage.sscan(
+                SET_WITH_SERVICES_MARKED_FOR_DELETION, cursor, count: SCAN_SLICE
+              )
+
+              res += services
+
+              break if cursor.to_i == 0
+
+              sleep(SLEEP_BETWEEN_SCANS)
+            end
+
+            res.to_set
+          end
+
+          def delete_keys(redis_conn, services, log_deleted_keys)
+            cursor = 0
+
+            loop do
+              cursor, keys = redis_conn.scan(cursor, count: SCAN_SLICE)
+
+              to_delete = keys.select { |key| delete_key?(key, services) }
+
+              unless to_delete.empty?
```

Comment:
You can mix this with the above with `tap do |to_delete|` (or a few other options) and remove the `to_delete` assignment.

Reply:
In my opinion, it comes down to personal preference. I don't see any obvious reason why any of those 2 approaches is better than the other.

Reply:
Less cognitive load and lower complexity. One assignment and a conditional (negated, slightly more complex) vs enum chaining with fewer lines and no conditionals. The error you made, not being sure the collection was or not empty is an indication that this is more error prone, just like writing manual loops vs using a functional style.

That said, up to you, no big deal.

Reply:
The conditional is also needed when using `tap`